### ISSUE TITLE
FastAPI ignores json_encoders in the Model

### ISSUE THREAD

**Issue by mikegrima** [tiangolo/fastapi#2091]
### First check

* [*] I added a very descriptive title to this issue.
* [*] I used the GitHub search to find a similar issue and didn't find it.
* [*] I searched the FastAPI documentation, with the integrated search.
* [*] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [*] I already read and followed all the tutorial in the docs and didn't find an answer.
* [*] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [*] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [*] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [*] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Below is an example that showcases the issue:

```Python
from typing import *

from pydantic import BaseModel
from fastapi.encoders import jsonable_encoder

def sort_func(x: Set[str]) -> List[str]:
   """This function should be called on every serialization. This will convert a Set of strings to a sorted List of strings."""
    print("i was called!")
    return sorted(x)


class SomeModel(BaseModel):
    stuff: Set[str] = ...

    class Config:
        """Define a custom encoder for Sets such that they are sorted lists."""
        json_encoders = {
            set: sort_func
        }

print(SomeModel(stuff={"z", "a", "f", "d"}).json())
# This prints:
# "i was called!"  # GOOD
# '{"stuff": ["a", "d", "f", "z"]}'  # GOOD
# The Pydandic serializer does what it's supposed to do.

print(jsonable_encoder(SomeModel(stuff={"z", "a", "f", "d"})))
# This prints:
# {'stuff': ['z', 'd', 'f', 'a']}  # BAD
# The fastapi encoder is not calling the `sort_func`
```

### Description

* The primary issue is that when a Pydantic model defines special serialization behavior, fastapi is not obeying it.
* In my example, I'm trying to store a set of values internally, but then serialize it into a sorted list.
* The desired behavior is for fastapi to be consistent and obey Pydantic's conventions.

I also looked at other issues that are similar, but not quite the same:
- https://github.com/tiangolo/fastapi/issues/1722
- https://github.com/tiangolo/fastapi/issues/1986

### Environment

* OS: macOS
* FastAPI Version: fastapi==0.61.1
* Python version: 3.8.5

### Additional context

I'm not sure if this is related but I think this is also a bug in the `jsonable_encoder`:
```Python
def jsonable_encoder(
    obj: Any,
    include: Optional[Union[SetIntStr, DictIntStrAny]] = None,
    exclude: Optional[Union[SetIntStr, DictIntStrAny]] = None,
    by_alias: bool = True,
    exclude_unset: bool = False,
    exclude_defaults: bool = False,
    exclude_none: bool = False,
    custom_encoder: dict = {},  # <------- 
    sqlalchemy_safe: bool = True,
) -> Any:
```
https://github.com/tiangolo/fastapi/blob/0dfde6e284b221bf6695a98762a56d13a995e807/fastapi/encoders.py#L34
The `custom_encoder` is a keyword argument that defines a mutable type.

**Comment by mikegrima** [comment]
I also tried overriding the `.json()` function (with the excellent suggestion from @reallistic). However, fastapi's `jsonable_encoder` doesn't call that either.

**Comment by ycd** [comment]
Hey @mikegrima, #2061 fixes this.You can temporarily create a workaround by moving custom encoder to the top.

https://github.com/tiangolo/fastapi/blob/e77ea635777d2494690ba3eb62bd005b9edeefde/fastapi/encoders.py#L116-L122

**Comment by mikegrima** [comment]
@ycd Thank you! I have confirmed that it does fix my bug. Would you like me to close this issue or keep it open until the PR is merged?

**Comment by ycd** [comment]
You are welcome, i think it should be open till it get's merged. but i did not understand, can you elaborate a little more ?

**Comment by mikegrima** [comment]
Sure, for the second part, take the `SomeModel` object:
```python
class SomeModel(BaseModel):
    stuff: Set[str] = ...
   
    def json(*args, **kwargs):
         # MY LOGIC HERE
         return json_str
```

If you override the `json` function, in Pydantic, it would call it when generating a JSON string. However, that does not work in FastAPI as it's not called. Is there some good guidance on the best way to override the JSON that is dumped? I'm very familiar with Marshmallow in that you can add functions to the object that are run on serialization after all the fields are processed, but not quite sure how to do that with Pydantic+FastAPI.

**Comment by mvanderlee** [comment]
@mikegrima FastAPI calls the `dict` function on Pydantic models, just like `json` does internally.

https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L65

**Comment by kklingenberg** [comment]
I'm having troubles with this fragment here:
https://github.com/tiangolo/fastapi/blob/fdb6c9ccc504f90afd0fbcec53f3ea0bfebc261a/fastapi/encoders.py#L43-L46

In my case, I'm attempting to temporarily apply encoding logic for a pydantic object but I very much don't want it to be permanent on the model's `Config.json_encoders` dict. If line 44 would instead be something like:

```python
encoder = dict(getattr(obj.__config__, "json_encoders", {}))
```

So as to copy the original encoders dict, so the permanent config mutation doesn't occur. I'm not sure this is a bug strictly, just an expectation issue: the `jsonable_encoder` should not ever mutate a pydantic class. Then again, I think the `custom_encoder` parameter is not even documented and maybe it's not intended for general use.

**Comment by mikegrima** [comment]
I'm now trying to figure out the best way to get a `sort_keys=True` somehow passed into the JSON generation. I'm unsure of the best way to do that.

**Comment by mikegrima** [comment]
To help close the loop, if you wanted to have sorted JSON results back, you would need to have something that looks like this:

```python
import json
import typing
from starlette.responses import Response
from fastapi import APIRouter
from pydantic import BaseModel

class SortedJSONResponse(Response):
    media_type = "application/json"

    def render(self, content: typing.Any) -> bytes:
        return json.dumps(
            content,
            ensure_ascii=False,
            allow_nan=False,
            indent=None,
            separators=(",", ":"),
            sort_keys=True   # <---
        ).encode("utf-8")

class SomeModel(BaseModel):
   some_value: int
   some_other_value: str
   and_something_else: bool

some_router = APIRouter()

@some_router.get("/some/api/path", response_class=SortedJSONResponse)
async def get_sorted_json_response_api() -> SomeModel:
   return SomeModel(some_value: 1, some_other_value: "some other value", and_something_else: True)
```

**Comment by mdgilene** [comment]
> I'm having troubles with this fragment here:
> 
> https://github.com/tiangolo/fastapi/blob/fdb6c9ccc504f90afd0fbcec53f3ea0bfebc261a/fastapi/encoders.py#L43-L46
> 
> In my case, I'm attempting to temporarily apply encoding logic for a pydantic object but I very much don't want it to be permanent on the model's `Config.json_encoders` dict. If line 44 would instead be something like:
> 
> ```python
> encoder = dict(getattr(obj.__config__, "json_encoders", {}))
> ```
> 
> So as to copy the original encoders dict, so the permanent config mutation doesn't occur. I'm not sure this is a bug strictly, just an expectation issue: the `jsonable_encoder` should not ever mutate a pydantic class. Then again, I think the `custom_encoder` parameter is not even documented and maybe it's not intended for general use.
> 
> EDIT: I'll leave my current workaround here: instead of passing a pydantic object to `jsonable_encoder`, I call the object's `dict()` method beforehand:
> 
> ```python
> # assume obj is a pydantic object
> # assume custom_encoder contains the custom encoders to be given for this encoding only
> encoded = jsonable_encoder(obj.dict(), custom_encoder=dict(obj.__config__.json_encoders, **custom_encoder))
> ```

I just ran into this exact same issue. `jsonable_encoder` modifying the classes config has to be a bug. If not, this is EXTREMELY unintuitive behavior and should be changed.

**Comment by KPull** [comment]
Sorry if this question was answered elsewhere but what is the reason that FastAPI does not call `json(..)` on the Pydantic `BaseModel`? Given that Pydantic already provides functionality for its models to be serialized as JSON, shouldn't FastAPI simply delegate to the model?

**Comment by TheBubblePopped** [comment]
+1 for this merge.

**Comment by gwerbin** [comment]
What's the workaround for the original issue? Overriding the model `.dict` method is pretty hacky because it affects *other* usages of the model. It seems like the only solution is to manually invoke `jsonable_encoder` yourself and mess around with the result until it looks the way you want. Big big +1 for making it possible to customize JSON serialization *somehow* whether or not using Pydantic config.

**Comment by takaria0** [comment]
Agreed. In my case I wanted to change the parsing of `datetime.time` behavior but couldn't find an elegant solution. Just like this `10:30:00` -> `1030` .

At first I thought adding `Config` in `BaseModel` would solve this problem, but it was not working. 

```python
    class Config:
        """Define a custom encoder for Sets such that they are sorted lists."""
        json_encoders = {
            datetime.time: some_func
        }
```

I also think adding `json_encoders` in `Config` is consistent with the current behavior, since we are already using `use_orm: bool` in the `Config`, right?
